### ISSUE TITLE
[1.2+] Full ES2015 module support is claimed, but strict mode not on

### ISSUE THREAD

**Issue by AnthonyAstige** [meteor/meteor#7050]
**The claims**
[The Meteor Migration Guide](http://guide.meteor.com/1.3-migration.html#modules) states,
> The biggest new feature in Meteor 1.3 is support for ES2015 modules on the client and the server.

[The Meteor Guide](http://guide.meteor.com/structure.html#es2015-modules) states (**emphasis** mine),
> As of version 1.3, Meteor ships with **full** support for ES2015 modules

**The problem**
[ES2015 Modules imply strict mode](https://stackoverflow.com/questions/29283935/which-ecmascript-6-features-imply-strict-mode), however [this repository confirms modules are not in strict mode](https://github.com/AnthonyAstige/meteor-modular-strict-test).

**Related**
[ESLint is not happy with this disjuncture](https://forums.meteor.com/t/eslint-for-meteor-1-3-modules-and-strict-mode/23105)

**Comment by laosb** [comment]
That's true. Hope someone can fix this.

**Comment by AnthonyAstige** [comment]
I'm guessing it's in the [meteor/ecmascript](https://github.com/meteor/meteor/tree/devel/packages/ecmascript) module, which [uses Babel](https://github.com/meteor/meteor/tree/devel/packages/ecmascript#syntax).  And it appears [Babel strict mode is an option](https://stackoverflow.com/questions/33821312/how-to-remove-global-use-strict-added-by-babel).

The documented [list of Babel modules](https://github.com/meteor/meteor/tree/devel/packages/ecmascript#syntax) doesn't include [babel-plugin-transform-strict-mode](https://babeljs.io/docs/plugins/transform-strict-mode/), and it might be simple as enabling that.

I do wonder however, if strict mode it enabled, what might break.  Both internally in Meteor, and for the community building projects on Meteor. This may have to be done on a major release (Meteor 2.0?), and/or set as an optional configuration/flag defaulting to false.

**Comment by abernix** [comment]
Thanks for the clear explanation and reproduction.

This is done on purpose right now to maintain the highest level of reverse compatibility with pre-1.3 versions.  See related issue here: #6166 and commits ae90250f80bc705f46c81d8a97ce252a77b0145c & https://github.com/meteor/babel/commit/823ac5721e5af4097f7f52ee7d5e4806536b5fc5 (which intentionally turned it off) and bugs here #6072 #6086 (not to mention all the other stuff that could be broken if this was changed right now).

If you want to use strict, you can (as you can see in your example) use `use strict` yourself and it will work properly.  It's still safe to say that Meteor 1.3 has full support for ES6 modules, but it's default setting interprets things slightly more loosely.  Of course, using linting settings that do not allow you to use globals is encouraged and hopefully a future version of Meteor can change this entirely.

So this is intended behavior – for now.

**Comment by AnthonyAstige** [comment]
@abernix - Thanks for the detailed reply! It's good to see examples why just turning on strict mode would break things.

I'm [contemplating adding to eslint-plugin-meteor](https://github.com/dferber90/eslint-plugin-meteor/issues/208), but that still leaves having to place `'use strict;'` on the top of potentially hundreds of  `/import/**.js` files

Before I delve into making lint rule, I wonder if it's possible to create an npm package to address this. It'd be nicer to do something like `meteor npm install meteor-strict-imports --save`. Is there anything in the Meteor core which a package could hook into to achieve this?

**Comment by abernix** [comment]
>It'd be nicer to do something like meteor npm install meteor-strict-imports --save. Is there anything in the Meteor core which a package could hook into to achieve this?

Great question! I'm not sure and nothing jumped out at me at first glance.  You'd definitely be potentially risking breaking some third-party packages but this is provided internally by the `meteor-babel` NPM package, but I'm not sure installing a local copy is possible (my very quick attempt said "no").

I'll leave this here and see if someone else can comment and hopefully find an answer for you within a couple days.

(Worst case scenario, if you're using SublimeText (or another regex-capable editor), I would do a global-**ish** RegEx search-and-replace with the "Find" as `(?-m)^` and the "Replace" as `'use strict';\n\n` which should add it to the top of all your files 😁 )

**Comment by AnthonyAstige** [comment]
>  You'd definitely be potentially risking breaking some third-party packages 

I'm hoping there's a way to just apply strict mode to modules in the `/imports/**.js` and leave the Meteor core safe.  I can trust the Meteor core to work, but I can't trust my own code and I want to enforce strict mode there to help prevent bugs.

> search-and-replace

Yeah I have tools for that (I'm working old school with vim & shell scripts).  I actually already have `'use strict';` on the top of my files from when I was putting all my code in separate packages before Meteor 1.2 allowed for import/export isolation.  The trouble remains though that as new files get created it can be forgotten.  Before I was using import/export my linter always caught it, now ESLint (correctly to ES2015 standards) says to not put strict mode on top since it's implicit to modules.

Anyways we'll see if someone can chime in with a package solution in a few days, otherwise the custom eslint rule may be the best bet.  Thanks again!

**Comment by abernix** [comment]
@AnthonyAstige Before I ping anyone else into this, I'm pretty sure PR #6874 might actually cover this?  Please also check conversation in #6351 and (the badly labeled) #6028.  Let me know what you think.

**Comment by AnthonyAstige** [comment]
@abernix 
Configurable babel plugins could be a great bandaid for this issue. Sounds like it can be done now without breaking changes; looks like the aim is to apply extra babel plugins optionally to project code and presumably never to meteor core code.

This issue may still apply to Meteor 2.0, but if a configurable babel plugin feature comes soon there's probably no need to build a custom lint rule (better to go with the broader solution to help other people too if it's not too hard to implement).

I see #6351 is set for the [Release 1.3.3](https://github.com/meteor/meteor/milestones/Release%201.3.3) milestone.  Does this mean a solution is likely to be included in Meteor 1.3.3?

**Comment by abernix** [comment]
I'll have to defer to @benjamn.

Ben, can the change in https://github.com/meteor/babel/commit/823ac5721e5af4097f7f52ee7d5e4806536b5fc5 (disabling strict mode in ES6 modules) be overridden by the end-user in any way right now?  Or, alternatively, if `.babelrc` support floats in some form (#6351), could this preference be achieved through that? (For end-user stuff in `/imports`, not Meteor as a whole)

**Comment by abernix** [comment]
@AnthonyAstige I think this is good news for you.  Could you check this out?  I'm too busy to test right now but would appreciate your confirmation on it.

The `release-1.3.3` branch got this: https://github.com/meteor/meteor/commit/7c63a8e704e73405eaa32042c997da957891e8b2 which *should* mean that if you:
* Setup your `package.json` [like this](https://github.com/meteor/meteor/blob/7c63a8e704e73405eaa32042c997da957891e8b2/tools/tests/apps/modules/packages/modules-test-plugin/package.json#L6) (you can skip the class-transform plugin)
* and `meteor npm install babel-plugin-transform-strict-mode`

...you might get what you want. *fingers crossed*

**Comment by abernix** [comment]
@AnthonyAstige it might have just changed a bit – [see here](https://github.com/meteor/meteor/pull/7033#issuecomment-220664981) as it now supports `.babelrc`. ❗ :smile:  Same branch.

**Comment by AnthonyAstige** [comment]
@abernix That sounds great.  I want to try it on [meteor-modular-strict-test](https://github.com/AnthonyAstige/meteor-modular-strict-test) but both `meteor update --release 1.3.3` and `meteor update --release release-1.3.3` return `unknown release`.  What should I be doing to work directly from the [release-1.3.3 branch](https://github.com/meteor/meteor/tree/release-1.3.3)?

**Comment by abernix** [comment]
Sorry, forgot to mention – you'll have to run [from a checkout](https://github.com/meteor/meteor/blob/devel/README.md#slow-start-for-developers), using the `release-1.3.3` branch.  Basically... 
```sh
cd ~/your-dev-dir
git clone https://github.com/meteor/meteor.git
cd meteor
git fetch origin
git checkout -t origin/release-1.3.3
cd path/to/your/app
path/to/meteor/meteor run
```

**Comment by AnthonyAstige** [comment]
Awesome it works! Great solution with the .babelrc, and thanks for the pointers running from a checkout too. :)

My test repo w/documentation updated at https://github.com/AnthonyAstige/meteor-modular-strict-test

Your call if this issue should be considered resolved by 1.3.3. Though strict mode off isn't the behavior I would expect by default, it looks like I'll have enough to easily turn it on.

**Comment by abernix** [comment]
Great to hear!  A beta of 1.3.3 shipped last night with this included – you can check this out with `meteor update --release 1.3.3-beta.0`.

I'm going to close this side of the issue and open meteor/guide#457 with [the Meteor Guide repo](http://github.com/meteor/guide/) to make sure thi is more clear.  And as soon as Meteor is not trying to be super-backwards-compatible, this will work itself out.

Fixed by 49a60f155b64e00a5f3e879540ce4be7bbb74691
